### DISCUSSION TITLE
Concurrency problem

### DISCUSSION THREAD

**Discussion by BowgartField** [pmndrs/jotai#1246] [Q&A] [answered]
Hi, 
I have a atom value storing a dictionnary:
`export const validators = atom<{}>({})`

This atom is used in same `<MyComponent/>` created like that:
`XXX.map(x => <MyComponent x={x}/>` . So they are initialized at "same time".
I use this atom in useEffect() at init of `<MyComponent/>` and this create a concurency problem because they are trying to update the same atom at same time.
The result is that only last component created by map() function write his value in dictionnary in validators atom.
Do you have any advice ?
Thanks !

EDIT: Also receiving this error in console (But in don't know if it's related):
![image](https://user-images.githubusercontent.com/13118844/174762128-bb6044a8-ff3e-4d53-bcbc-ce98f8d518ab.png)

**Comment by dai-shi** [comment, marked as answer]
Can you show more of your code? How does the atom used in the component?
You can't update the atom value in render function.

**Reply by BowgartField** [reply]
Yes.
This is the atom:
`export const deployContentValidators = atom<{}>({})`

The Component (I simplified it):
![image](https://user-images.githubusercontent.com/13118844/174769699-e8935bdc-29eb-4b27-a03c-5e7a6ad31c87.png)

**Reply by dai-shi** [reply]
Can you try this?
```js
const setValidator = async (value) => {
  // ...
  await setValidators((prev) => ({ ...prev, [fieldId]: value }))
}
```

I think it's a general React gotcha. It can happen with `useState` too.
The eslint rule should warn you missing `setValidator` in useEffect deps.

**Reply by BowgartField** [reply]
Thanks that works ! 
Yes elsint warn me but if i add setValidor it creates a infinit rendering.

**Reply by dai-shi** [reply]
You can use `useCallback` to avoid the infinite loop, now because it doesn't depend on `validators` variable.
